sources/lib/api/gcp/bigquery/_query.py

```diff
@@ -66,23 +68,74 @@ def __init__(self, sql, scripts=None, context=None, values=None, **kwargs):
     Raises:
       Exception if expansion of any variables failed.
       """
-    if kwargs or values or not isinstance(sql, basestring):
-      if values is None:
-        values = kwargs
-      sql, code = gcp.data.SqlModule.expand(sql, values)
-      if code:
-        if scripts is None:
-          scripts = code
-        else:
-          scripts.extend(code)
-
     if context is None:
       context = gcp.Context.default()
     self._context = context
     self._api = _api.Api(context)
     self._sql = sql
-    self._scripts = scripts
     self._results = None
+    self._code = None
+    self._imports = []
+    if values is None:
+      values = kwargs
+
+    self._sql = gcp.data.SqlModule.expand(sql, values, udfs)
+
+    # We need to take care not to include the same UDF code twice so we use sets.
+    udfs = set(udfs if udfs else [])
+    for value in values.values():
+      if isinstance(value, _udf.UDF):
+        udfs.add(value)
+    included_udfs = set([])
+
+    tokens = gcp.data.tokenize(self._sql)
+    udf_dict = {udf.name: udf for udf in udfs}
+
+    for i, token in enumerate(tokens):
+      prior = i - 1
+      while prior >= 0 and tokens[prior].isspace():
+        prior -= 1
+      if prior < 0 or not tokens[prior].upper() == 'FROM':
+        continue
+      next = i + 1
+      while next < len(tokens) and tokens[next].isspace():
+        next += 1
+      if next >= len(tokens) or not tokens[next] == '(':
+        continue
+
+      # We know now we have a 'FROM token (' sequence.
+
+      if token not in udf_dict:
+        utoken = token.upper()
+        if utoken != 'TABLE_DATE_RANGE' and utoken != 'TABLE_DATE_RANGE_STRICT' and \
```

Comment:
Will it work if we don't check for this? By work, I mean BigQuery will raise an appropriate error message. It would be better if we didn't hardcode some select function names here.